### PR TITLE
Test upgrading from an older version

### LINKED ISSUE #13183: Packaging tests should verify upgrades from some previous version
Right now the bats tests install elasticsearch on a clean system. They should also test that the packages can be upgraded without losing data. Both across major and minor versions.

### PR DESCRIPTION
**Pull request by nik9000**
Adds a test for upgrading from 2.0.0-beta1 to the version of that is built.

Closes #13183

### CODE REVIEW

File: qa/vagrant/src/test/resources/packaging/scripts/30_deb_package.bats [hunk 1]
```diff
@@ -46,7 +46,7 @@ setup() {
 }
 
 @test "[DEB] package is available" {
-    count=$(find . -type f -name 'elastic*.deb' | wc -l)
+    count=$(ls elasticsearch-$(cat version).deb | wc -l)
```
Comment (dakrone) [inline qa/vagrant/src/test/resources/packaging/scripts/30_deb_package.bats]: Where/how is `version` set up?

File: qa/vagrant/src/test/resources/packaging/scripts/80_upgrade.bats [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+load packaging_test_utils
+
+# Cleans everything for the 1st execution
+setup() {
+    skip_not_dpkg_or_rpm
+}
+
+@test "[UPGRADE] install old version" {
+    clean_before_test
+    install_package -v $(cat upgrade_from_version)
```
Comment (dakrone) [inline qa/vagrant/src/test/resources/packaging/scripts/80_upgrade.bats]: Not clear where upgrade_from_version comes from?

File: qa/vagrant/src/test/resources/packaging/scripts/30_deb_package.bats (hunk 1 above)
Comment (nik9000) [inline qa/vagrant/src/test/resources/packaging/scripts/30_deb_package.bats]: In the pom line 240.

File: qa/vagrant/src/test/resources/packaging/scripts/80_upgrade.bats (hunk 2 above)
Comment (nik9000) [inline qa/vagrant/src/test/resources/packaging/scripts/80_upgrade.bats]: The pom line 242.

I can certainly add comments to these lines about the mysterious origins of these file.

File: qa/vagrant/src/test/resources/packaging/scripts/80_upgrade.bats [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+@test "[UPGRADE] install version under test" {
+    install_package -u
+}
+
+@test "[UPGRADE] start version under test" {
+    start_elasticsearch_service yellow
+}
+
+@test "[UPGRADE] check elasticsearch version is version under test" {
+    local versionToCheck=$(cat version | sed -e 's/-SNAPSHOT//')
```
Comment (dakrone) [inline qa/vagrant/src/test/resources/packaging/scripts/80_upgrade.bats]: hardcoding `-SNAPSHOT` seems dangerous here, what if we also need to strip something like `-RC` also?

File: qa/vagrant/src/test/resources/packaging/scripts/packaging_test_utils.bash [hunk 4]
```diff
@@ -318,6 +333,8 @@ clean_before_test() {
 }
 
 start_elasticsearch_service() {
+    local desiredStatus=${1:-green}
+
```
Comment (dakrone) [inline qa/vagrant/src/test/resources/packaging/scripts/packaging_test_utils.bash]: So this is getting kind of silly, but I think we may want to start "bashdoc"-ing these functions?

File: qa/vagrant/src/test/resources/packaging/scripts/packaging_test_utils.bash [hunk 5]
```diff
@@ -389,10 +406,7 @@ stop_elasticsearch_service() {
 
 # Waits for Elasticsearch to reach a given status (defaults to "green")
 wait_for_elasticsearch_status() {
-    local desired_status="green"
-    if [ "x$1" != "x" ]; then
-        status="$1"
-    fi
+    local desiredStatus=${1:-green}
```
Comment (dakrone) [inline qa/vagrant/src/test/resources/packaging/scripts/packaging_test_utils.bash]: Should we validate `$1` to be `green|yellow|red`?

File: qa/vagrant/src/test/resources/packaging/scripts/80_upgrade.bats (hunk 2 above)
Comment (dakrone) [inline qa/vagrant/src/test/resources/packaging/scripts/80_upgrade.bats]: That would be helpful for people like me :)

**Comment by dakrone** [comment]
The bash portions of this look good to me, the maven portions.... look like XML that does Magic™

File: qa/vagrant/src/test/resources/packaging/scripts/packaging_test_utils.bash (hunk 4 above)
Comment (nik9000) [inline qa/vagrant/src/test/resources/packaging/scripts/packaging_test_utils.bash]: Almost certainly.

File: qa/vagrant/src/test/resources/packaging/scripts/packaging_test_utils.bash (hunk 5 above)
Comment (nik9000) [inline qa/vagrant/src/test/resources/packaging/scripts/packaging_test_utils.bash]: Probably.

**Comment by jpountz** [comment]
LGTM. I'm fine with doing it with 2.0.0 beta1 for now in order to build the infra but since we don't guarantee anything in terms of compatibility of this version, I think we should move to 2.0.0 as soon as it's released? If you agree, maybe we should leave a comment about it?

**Comment by nik9000** [comment]
> LGTM. I'm fine with doing it with 2.0.0 beta1 for now in order to build the infra but since we don't guarantee anything in terms of compatibility of this version, I think we should move to 2.0.0 as soon as it's released? If you agree, maybe we should leave a comment about it?

Agreed. I'll leave a comment.

**Comment by nik9000** [comment]
Ok - pushed some extra docs like the reviewers asked for. OTOH they both said it looked good to them so I'm just going to merge this as well.

**Comment by nik9000** [comment]
Pushed to master, 2.x, and 2.0.

**Pull request merged**